### PR TITLE
Typography: Make title blocks apply typographic styles consistently

### LINKED ISSUE #65169: Site Title block conflicts between typography and link styles
https://github.com/WordPress/gutenberg/pull/64911 resolved the application of typography with blocks that have an `a` as a child, like the core/site-title block. 

In the case of the core/site-title block, any heading styles applied via theme.json will still effect the block's heading element. 

To replicate, in blocks: 

```
"core/site-title": {
                "elements": {
                    "link": {
                        "typography": {
                            "textDecoration": "none"
                        }
                    }
                },
                "typography": {
                    "fontSize": "var(--wp--preset--font-size--medium)",
                    "fontWeight": "550"
                }
            }
```

and in elements: 

```
"h1": {
                "typography": {
                    "fontSize": "var(--wp--preset--font-size--xx-large)",
                    "lineHeight": "1"
                }
            },
```

CleanShot 2024-09-09 at 16 26 25

Originally reported https://github.com/WordPress/gutenberg/pull/64911#issuecomment-2339018022.

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Fixes: https://github.com/WordPress/gutenberg/issues/65169

Addresses issues introduced by https://github.com/WordPress/gutenberg/pull/64911

## What?

When title related blocks (Site Title and Post Title) are set to contain a link, typographic styles from the wrapping heading are enforced on the inner link.

## Why?

See discussion here: https://github.com/WordPress/gutenberg/issues/65169#issuecomment-2345321749

TL;DR

- These blocks are conceptually headings regardless of the presence of an inner link
- Inner link styles could still be applied on a block instance or through the `elements` within a block type definition in theme.json
- Currently, the blocks have a `heading > link` structure and so global element styles for links and headings conflict
- Recent changes so that global styles for the block type get applied over global link element styles introduced a couple of bugs e.g. styles being inconsistently applied between editor and frontend.

## How?

- Remove custom typography selectors introduced in https://github.com/WordPress/gutenberg/pull/64911 (fixes inconsistent styling issue)
- Consolidates "inheritance" styles in block library to enforce typographic style inheritance on links within the title blocks.

## Testing Instructions

1. Activate emptytheme or remove `fontSize` from TT4's Site Title styles in theme.json.
2. Create a post with multiple Site Title and Post Title blocks. Make sure that each block type has one that is set to contain a link and one without.
3. Save the post and navigate to Appearance > Editor > Styles > Typography
4. Apply some link element styles, save, then confirm these do not apply to the Site Title and Post Title blocks in the previous post.
5. Back in Global Styles, apply some global heading element typography styles. This time confirm this does apply to the Site Title and Post Title blocks. 
6. Confirm the styling is consistently applied in both the editor and frontend.
7. Back in the site editor, navigate to Styles > Blocks. Apply typographic styles for both block types, then save.
8. Confirm that the global block type styles override the global heading element styles.

<details>
<summary>Example block markup for testing</summary>

```html
<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>Site Title with inner link</p>
<!-- /wp:paragraph -->

<!-- wp:site-title /--></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>Site Title without inner link</p>
<!-- /wp:paragraph -->

<!-- wp:site-title {"isLink":false} /--></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->

<!-- wp:separator {"className":"is-style-wide"} -->
<hr class="wp-block-separator has-alpha-channel-opacity is-style-wide"/>
<!-- /wp:separator -->

<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>Post Title with inner link</p>
<!-- /wp:paragraph -->

<!-- wp:post-title {"isLink":true} /--></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>Post Title without inner link</p>
<!-- /wp:paragraph -->

<!-- wp:post-title /--></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->

<!-- wp:separator {"className":"is-style-wide"} -->
<hr class="wp-block-separator has-alpha-channel-opacity is-style-wide"/>
<!-- /wp:separator -->

<!-- wp:heading -->
<h2 class="wp-block-heading">Heading</h2>
<!-- /wp:heading -->

<!-- wp:paragraph -->
<p>Paragraph with a <a href="http://www.wordpress.org">link</a></p>
<!-- /wp:paragraph -->
```

</details>

## Screenshots or screencast 

| Frontend Before | Frontend After |
|---|---|
| Screenshot 2024-09-13 at 3 32 17 PM | Screenshot 2024-09-13 at 3 36 16 PM |

Demo adjusting global link, heading element and block type styles

https://github.com/user-attachments/assets/e37e9250-80f6-493c-b1fd-11c540631038

### CODE REVIEW

**Comment by aaronrobertshaw** [comment]
I'm not sure whether this proposal would be considered a breaking change if themes relied on previous broken behaviour. I'm more than happy to hear any thoughts or concerns. 

Until this aspect of the PR reaches a consensus, I've left it in a draft state.

**Comment by andrewserong** [comment]
> Until this aspect of the PR reaches a consensus, I've left it in a draft state.

Good point! Thanks for getting this PR up quickly. I'll likely run out of time to give it a proper test today, but will take it for a spin on Monday.

**Review by andrewserong — COMMENTED** [review]
Thanks for the follow-up here @aaronrobertshaw, conceptually I think this is the right way to go, and it means that we're saying that these blocks effectively _are_ headings, rather than thinking of them as links. That greatly reduces the complexity of dealing with these blocks, IMO.

This is testing well for me:

✅ The blocks are no longer affected by overall link element styles
✅ Global heading element styles (e.g. letter spacing) apply to the Site Title and Post Title blocks where those blocks don't set those styles
✅ Block level styles in global styles override heading element styles as expected
✅ Block instance level styles continue to override block-level in global styles rules as expected

image

So, this is all looking good to me so far.

> Until this aspect of the PR reaches a consensus, I've left it in a draft state.

Would it be worth doing a wider ping to the design or outreach groups for feedback, just to get a sense of whether anyone thinks it's a risk switching to always inheriting from the heading's styles?

**Comment by aaronrobertshaw** [comment]
> Would it be worth doing a wider ping to the design or outreach groups for feedback, just to get a sense of whether anyone thinks it's a risk switching to always inheriting from the heading's styles?

Definitely, at least now its had a basic sanity check by a few people 👍 

I've given this PR a rebase so tests should pass soon. 

--- 

@WordPress/gutenberg-design and @WordPress/outreach it would be great to hear your thoughts on the approach in this PR 🙏 

Effectively, as it stands in trunk, we can't reconcile link (`a`) element, heading elements (`h1-h6`), and global block type styles. 

This PR explores treating Site Title and Post Title blocks as only "headings" i.e. even if they are set to contain a link they won't pick up global link element styles that might prevent other styles applying cleanly. See the PR description and linked issue for more context.

**Comment by justintadlock** [comment]
Haven't had a chance to fully test yet, but why add this to CSS files instead of the Core/Gutenberg default `theme.json`?

For testing instructions, I'd add something like this to test from TT4's `theme.json` to ensure it's working:

```json
{
	"styles": {
		"blocks" {
			"core/site-title": {
				"color": {
					"text": "blue"
				},
				"elements": {
					"link": {
						"color": {
							"text": "red"
						}
					}
				}
			}
		}
	}
}
```

Also test custom typography on both the heading wrapper and nested link.

**Comment by justintadlock** [comment]
My biggest worry here is that it's going to break the colors for some themes. General link colors should always apply unless there is more specific styling for the Post and Site Title nested links. Breaking these expectations while working on themes is annoying.

I'd avoid inheriting the color by default because:

- Theme authors may be relying on the cascade and their base link colors to apply.
- When a user sets a text color in the UI, WP automatically sets the link color too.
- It's beyond the scope of the ticket's proposal (as far as I can see, there's not one mention of the text color, but it's included in the code changes).

I feel like the typography changes wouldn't have much of an impact, but I may be missing some scenarios.

For me, though, I wanted to ensure that what I set in `theme.json` was overruling the CSS. As far as I can tell, customizing the colors and typography at the block and nested link element level worked. Testing code:

```json
{
	"$schema": "https://raw.githubusercontent.com/WordPress/gutenberg/trunk/schemas/json/theme.json",
	"version": 3,
	"styles": {
		"blocks": {
			"core/site-title": {
				"color": {
					"text": "purple"
				},
				"typography": {
					"fontSize": "19px",
					"fontStyle": "normal",
					"fontWeight": "500",
					"textTransform": "uppercase"
				},
				"elements": {
					"link": {
						"color": {
							"text": "red"
						},
						"typography": {
							"textDecoration": "none"
						},
						":focus": {
							"color": {
								"text": "blue"
							},
							"typography": {
								"textDecoration": "underline"
							}
						},
						":hover": {
							"color": {
								"text": "orange"
							},
							"typography": {
								"textDecoration": "underline"
							}
						}
					}
				}
			}
		}
	}
}
```

**Comment by aaronrobertshaw** [comment]
Thanks for the questions and calling out the color styles @justintadlock 👍 

> Haven't had a chance to fully test yet, but why add this to CSS files instead of the Core/Gutenberg default theme.json?

The CSS is there to handle block instance styles coming from block supports. These need to be present for classic themes as well which won't have access to the default theme.json and global styles stylesheet.

This PR consolidates those existing styles into a single rule.

> My biggest worry here is that it's going to break the colors for some themes.

The color inheritance for the Site Title block's link is [already there in trunk](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/site-title/style.scss?plain=1#L30). I did copy the Site Title's inheritance styles to the Post Title block and missed that it didn't have the color inheritance. 

That does raise a question though shouldn't these blocks be consistent? I can see that the use cases for the Post Title (within query loops etc) might be different than a Site Title so perhaps they shouldn't 🤔


> I'd avoid inheriting the color by default because:

I can remove the new `color: inherit` style from the Post Title block. Removing it from the Site Title would possibly be a breaking change though.

> there's not one mention of the text color

This was an oversight of mine once the styles settled. 

> For me, though, I wanted to ensure that what I set in theme.json was overruling the CSS. As far as I can tell, customizing the colors and typography at the block and nested link element level worked. Testing code:

That's correct. The style changes here only affect how global heading and link element styles apply by default. The global block type styles are enqueued after the library and directly target the block and inner elements as per the definition.

FWIW, I believe you can still recreate the original typography issue by defining font-size and line-height on one of these blocks and then also defining the same typography styles on the inner link. In my view, that's a case of "doing it wrong" though we don't need to optimize for.

--- 

I'll push a tweak removing the color inheritance for the Post Title block shortly. Let me know what you think of the end result 🙏

**Comment by justintadlock** [comment]
> That does raise a question though shouldn't these blocks be consistent? I can see that the use cases for the Post Title (within query loops etc) might be different than a Site Title so perhaps they shouldn't 🤔
>
> I can remove the new `color: inherit` style from the Post Title block. Removing it from the Site Title would possibly be a breaking change though.

Yes, I was more thinking about post titles with color changes. The site title doesn't need to change. I don't think consistency is important here because post titles have so many more use cases.

Update looks good to me. I'm not seeing any breakage across a few themes.

**Review by vcanales — APPROVED** [review]
+1 regarding this working as expected and fixing the referenced issues.

I think this is a good move as expressed in https://github.com/WordPress/gutenberg/issues/65169, and it looks like we're only hoping for a sanity check on the move to treating Site and Post Title blocks as headings. To that effect, it might be a good idea to ship this with 19.3 RC, due tomorrow, to get some actual usage and exposure. From there, we could contrast our concerns regarding breakage with some real-life usage.

**Comment by aaronrobertshaw** [comment]
Thank you everyone for the reviews and testing 🙇 

> To that effect, it might be a good idea to ship this with 19.3 RC, due tomorrow, to get some actual usage and exposure.

Agreed. I'll get this merged.

Happy to spin up any necessary follow-ups when the time comes.
